dbms/src/Storages/KVStore/ProxyStateMachine.h

```diff
@@ -0,0 +1,442 @@
+        , log(log_)
+    {}
+
+    void join() const
+    {
+        if (!parms.conf.is_proxy_runnable)
+            return;
+        pthread_join(thread, nullptr);
+    }
+
+    void run()
+    {
+        if (!parms.conf.is_proxy_runnable)
+            return;
+        pthread_attr_t attribute;
+        pthread_attr_init(&attribute);
+        pthread_attr_setstacksize(&attribute, parms.stack_size);
+        LOG_INFO(log, "Start raft store proxy. Args: {}", parms.conf.args);
+        pthread_create(&thread, &attribute, runRaftStoreProxyFFI, &parms);
+        pthread_attr_destroy(&attribute);
+    }
+
+private:
+    static void * runRaftStoreProxyFFI(void * pv)
+    {
+        setThreadName("RaftStoreProxy");
+        const auto & parms = *static_cast<const RunRaftStoreProxyParms *>(pv);
+        run_raftstore_proxy_ffi(static_cast<int>(parms.conf.args.size()), parms.conf.args.data(), parms.helper);
+        return nullptr;
+    }
+
+    RunRaftStoreProxyParms parms;
+    pthread_t thread{};
+    const LoggerPtr & log;
+};
+
+struct ProxyStateMachine
+{
+    ProxyStateMachine(LoggerPtr log_, TiFlashProxyConfig && proxy_conf_)
+        : log(std::move(log_))
+        , proxy_conf(std::move(proxy_conf_))
+    {
+        helper = GetEngineStoreServerHelper(&tiflash_instance_wrap);
+        proxy_runner = std::make_unique<RaftStoreProxyRunner>(
+            RaftStoreProxyRunner::RunRaftStoreProxyParms{&helper, proxy_conf},
+            log);
+    }
+
+
+    // A TikvServer will be bootstrapped, FFI mechanism is enabled.
+    // However, the raftstore service is not started until we call `startProxyService`.
+    void runProxy()
+    {
+        if (proxy_conf.is_proxy_runnable)
+        {
+            proxy_runner->run();
+
+            LOG_INFO(log, "wait for tiflash proxy initializing");
+            while (!tiflash_instance_wrap.proxy_helper)
+                std::this_thread::sleep_for(std::chrono::milliseconds(200));
+            LOG_INFO(log, "tiflash proxy is initialized");
+        }
+        else
+        {
+            LOG_WARNING(log, "Skipped initialize TiFlash Proxy");
+        }
+    }
+
+    void initKVStore(TMTContext & tmt_context, std::optional<raft_serverpb::StoreIdent> & store_ident)
+    {
+        if (store_ident)
+        {
+            // Many service would depends on `store_id` when disagg is enabled.
+            // setup the store_id restored from store_ident ASAP
+            // FIXME: (bootstrap) we should bootstrap the tiflash node more early!
+            auto kvstore = tmt_context.getKVStore();
+            metapb::Store store_meta;
+            store_meta.set_id(store_ident->store_id());
+            store_meta.set_node_state(metapb::NodeState::Preparing);
+            kvstore->setStore(store_meta);
+        }
+        else
+        {
+            // Could be a auto-scaled compute node.
+            LOG_INFO(log, "KVStore is not initialized because no store_ident is provided");
+        }
```

Comment:
I think we can remove this branch. Now for non-disagg mode, the UniPS is not enabled so the `store_ident` is nullopt.

And for non-disagg mode, the StoreIdnet will be set by the FFI `fn_set_store`. Outputing this logging could be misleading.

Reply:
removed